### ISSUE TITLE
Move continuous adjoint and discrete VJP code together

### ISSUE THREAD

**Issue by facusapienza21** [ODINN-SciML/ODINN.jl#295]
Right now we have the continuous VJP method implemented in ODINN and the discrete VJP in Huginn. I think all should be in the same space. I think it makes more sense to have all in ODINN since Huginn is just for forward simulation, but I think both options are fine. 

What do you think @albangossard ? I can take care of this, but just want to check with you what do you think before moving forward!

**Comment by facusapienza21** [comment]
Following my comment in Slack, I think it makes more sence to move things to ODINN (and not Huginn) since the type of inversion will determine some internals of the VJPs, which right now live in Huginn. 

By doing this, I will also move the tests of the VJPs from Huginn to ODINN. 

Let me know if you approve this @albangossard and then I will move forward with a PR in the following days!

**Comment by albangossard** [comment]
Yes we should definitely standardize things!

I understand your opinion about moving the VJPs to ODINN but I think that from a software development point of view, it's much better to move the VJPs to Huginn for two reasons:
- Everything related to the iceflow should be in Huginn and this includes the VJPs although they are not really used in this package.
- If we move the VJPs to ODINN and we make some changes in the SIA2D, we won't see that we break the consistency between the forward and the VJP in the CI until we make a release of Huginn.

I think that the VJPs shouldn't depend on the neural network and it should only compute VJPs with respect to the input of SIA2D, that is with respect to H, A, D, etc. This is actually what was implemented for the discrete VJP and this allows to check the consistency without needing to define a neural network. Additionally, this makes the finite difference test faster for A since we only need to differentiate a scalar.

What is your opinion about this ? I'm happy to discuss how we could standardize the way we test VJPs.

**Comment by facusapienza21** [comment]
Thank you for the comments @albangossard ! I partially disagree with the first argument, but I fully agree with your second argument. **However, the VJPs will depend of the neural network.** For the simple inversion of A we where working until now, the dependency of the VJP between the physical parameters (H, A, D) and the NN (\theta) was separable, as you wrote in your code where the \nabla \theta is applied after VJPs. However, in more complicated cases this won't be longer the case. A good example is when the whole D is replaced by a NN that takes H as argument. 

I came to this conclusion while working on #298. I needed to massage the VJP code (see the discrete) to new target designs (althought is very nice to see most of the structure of the code does't change, since it is related to the structure of the PDE).  If you see the `Target()` type (which is defined for the UDE and then inside ODINN), this provides the information of the design of the inversion that needs to be feed in the VJPs (which if they are in Huginn, we need to pass somehow the target object)

I see both options can work, and I am happy to keep thinking about this (maybe a good thing to discuss in a conversation).

**Comment by albangossard** [comment]
Well for the particular case where the diffusivity depends on H, we can still split the VJP between a part that depends on the neural network, and a part that derives from SIA2D but which is not really the SIA. We could express this as:
$dH=f(H, simulation, D)$ where $D=g(H,\theta)$. This allows to handle the VJPs related to the physical equations in Huginn, and we only need to combine them with the neural network components in ODINN. Obviously this means that we need to treat the different cases depending on the combination of neural networks we want to use (w/ or w/o A, same thing for D, etc.) but I guess this problem also appears if we code all VJP related functions in ODINN.

The only problem I see is when the neural network isn't at the beginning or the end of the computational graph of SIA2D, and that's probably what you mean when you said
> in more complicated cases this won't be longer the case

but I'm not sure we will encounter such situations (again the diffusivity can still be separated). We could also code the different VJPs that represent the parts of the computational graph before and after the neural network in SIA2D directly in Huginn and then we could combine them in ODINN with the VJP of the neural network in between.

Anyway I think we should have this discussion with @JordiBolibar during our next meeting because this is critical for the structure of the packages.

**Comment by facusapienza21** [comment]
This is a great discussion to have in person! And I agree we should agree on this since it is important for the structure of the packages. 

I still don't think you can separate the case where the diffusivity depends on H: you have to evaluate the NN for different inputs in each spatial pixel, an operation that is handled inside the VJP computation (the sum that goes to the loss function). There are basic operations inside the VJP that will require the evaluation of the NN at different pixels. 

Furthermore, conceptually the adjoint/VJP strongly depends on the output of your computer program, which is given by the loss function and the design of what the parameters \theta is. This means that naturally VJPs need to be design knowing what the end point of the algorithm will be to run in reverse. Since VJPs are in reverse mode, you naturally need to have specified the whole model before doing this, meaning the model run in inverse, meaning it should belong to ODINN. 

Happy to keep talking about this!
